In [79]:
# Import required libraries and dependencies
import pandas as pd
import numpy as np
from pathlib import Path
import hvplot.pandas

from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

from sklearn.metrics import confusion_matrix, classification_report
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import GridSearchCV

import numpy as np
from scipy.stats import linregress

import matplotlib.pyplot as plt
import hvplot.pandas

In [2]:
### import and read csv file

path = 'Output/full_indicators_by_country_by_year.csv'
full_indicators_df = pd.read_csv(path)
# full_indicators = full_indicators.rename(columns={'total':'Immigration_flow'})

full_indicators_df.head(5)

index      country iso3Code  year  \
0      0  Afghanistan      AFG  2015   
1      1  Afghanistan      AFG  2016   
2      2  Afghanistan      AFG  2017   
3      3  Afghanistan      AFG  2018   
4      4  Afghanistan      AFG  2019   

   clean fuels for cooking (% of population)  \
0                                       27.6   
1                                       28.8   
2                                       30.3   
3                                       31.4   
4                                       32.6   

   Access to electricity (% of population)  Agricultural land (% )  \
0                                     71.5               58.123668   
1                                     97.7               58.123668   
2                                     97.7               58.123668   
3                                     93.4               58.276988   
4                                     97.7               58.276988   

   Armed forces (% labor force)  precipitation (mm per year)  \
0                      3.724667                        327.0   
1                      3.635915                        327.0   
2                      3.501771                        327.0   
3                      2.961201                        327.0   
4                      3.023536                        327.0   

   Birth rate, crude (per 1,000 people)  ...  \
0                                38.803  ...   
1                                37.936  ...   
2                                37.342  ...   
3                                36.927  ...   
4                                36.466  ...   

   Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)  \
0                                              0.009                                                      
1                                              0.013                                                      
2                                              0.017                                                      
3                                              0.016                                                      
4                                              0.016                                                      

   Total debt service (% of GNI)  \
0                       0.292137   
1                       0.324903   
2                       0.333475   
3                       0.344999   
4                       0.291267   

   Total debt service (% of exports of goods, services and primary income)  \
0                                           3.299608                         
1                                           4.333403                         
2                                           4.353572                         
3                                           3.323800                         
4                                           2.974238                         

   Unemployment advanced education  Unemployment basic education   \
0                             0.00                          0.000   
1                             0.00                          0.000   
2                            15.46                         12.283   
3                            15.46                         12.283   
4                            15.46                         12.283   

   Unemployment intermediate education  \
0                                0.000   
1                                0.000   
2                               16.155   
3                               16.155   
4                               16.155   

   Unemployment, total (% of total labor force) (modeled ILO estimate)  \
0                                              9.052                     
1                                             10.133                     
2                                             11.184                     
3                                             11.196                     
4          

In [3]:
# Generate summary statistics
full_indicators_df.describe()

index         year  clean fuels for cooking (% of population)  \
count  1663.000000  1663.000000                                1663.000000   
mean    831.000000  2019.034275                                  65.258328   
std     480.211065     2.589557                                  39.562522   
min       0.000000  2015.000000                                   0.000000   
25%     415.500000  2017.000000                                  25.600000   
50%     831.000000  2019.000000                                  87.500000   
75%    1246.500000  2021.000000                                 100.000000   
max    1662.000000  2023.000000                                 100.000000   

       Access to electricity (% of population)  Agricultural land (% )  \
count                              1663.000000             1663.000000   
mean                                 83.683824               38.149248   
std                                  27.096966               22.524579   
min                                   0.000000                0.000000   
25%                                  73.750000               19.602978   
50%                                  99.900000               39.434768   
75%                                 100.000000               54.637271   
max                                 100.000000               84.563658   

       Armed forces (% labor force)  precipitation (mm per year)  \
count                   1663.000000                  1663.000000   
mean                       1.140878                  1080.260313   
std                        1.549332                   803.647910   
min                        0.000000                     0.000000   
25%                        0.336253                   460.000000   
50%                        0.682330                   900.000000   
75%                        1.342953                  1622.000000   
max                       14.208087                  3240.000000   

       Birth rate, crude (per 1,000 people)  \
count                           1663.000000   
mean                              19.438815   
std                               10.131943   
min                                0.000000   
25%                               10.800000   
50%                               16.833000   
75%                               27.179000   
max                               46.558000   

       Central government debt, total (% of GDP)  \
count                                1663.000000   
mean                                   23.912632   
std                                    39.928103   
min                                     0.000000   
25%                                     0.000000   
50%                                     0.000000   
75%                                    44.000660   
max                                   249.411354   

       Consumer price index (2010 = 100)  ...  \
count                        1663.000000  ...   
mean                          315.239622  ...   
std                          1857.934281  ...   
min                             0.000000  ...   
25%                           112.430438  ...   
50%                           127.508994  ...   
75%                           159.761076  ...   
max                         38796.557297  ...   

       Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)  \
count                                        1663.000000                                                      
mean                                            5.118244                                                      
std                                             4.046082                                                      
min                                             0.000000                                                      
25%                                             1.420000                                                      
50%                          

---

### Prepare the Data

In [4]:
full_indicators_df.columns

Index(['index', 'country', 'iso3Code', 'year',
       'clean fuels for cooking (% of population)',
       'Access to electricity (% of population)', 'Agricultural land (% )',
       'Armed forces (% labor force)', 'precipitation (mm per year)',
       'Birth rate, crude (per 1,000 people)',
       'Central government debt, total (% of GDP)',
       'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
       'Cost of business start-up (% of GNI per capita)',
       'Current account balance (% of GDP)',
       'Current health expenditure (% of GDP)',
       'Death rate, crude (per 1,000 people)',
       'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
       'doing business score', 'Fertility rate, total (births per woman)',
       '% Firms using banks', 'GDP growth (annual %)',
       'GDP per capita (constant 2015 US$)',
       'GDP per capita, PPP (constant 2021 international $)',
       'GNI per capita (2015 US$)', 'government 

### Normalize the data using StandardScaler()

In [6]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file

indicators_scaled = StandardScaler().fit_transform(full_indicators_df[[

                                                                      'clean fuels for cooking (% of population)',
                                                                      'Access to electricity (% of population)', 'Agricultural land (% )',
                                                                      'Armed forces (% labor force)', 'precipitation (mm per year)',
                                                                      'Birth rate, crude (per 1,000 people)',
                                                                      'Central government debt, total (% of GDP)',
                                                                      'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
                                                                      'Cost of business start-up (% of GNI per capita)',
                                                                      'Current account balance (% of GDP)',
                                                                      'Current health expenditure (% of GDP)',
                                                                      'Death rate, crude (per 1,000 people)',
                                                                      'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
                                                                      'doing business score', 'Fertility rate, total (births per woman)',
                                                                      '% Firms using banks', 'GDP growth (annual %)',
                                                                      'GDP per capita (constant 2015 US$)',
                                                                      'GDP per capita, PPP (constant 2021 international $)',
                                                                      'GNI per capita (2015 US$)', 'government  consumption exp (% of GDP)',
                                                                      'Government expenditure on education, total (% of GDP)',
                                                                      'Human Capital Index',
                                                                      'Industry (including construction), value added (% of GDP)',
                                                                      'Inflation, consumer prices (annual %)', 'Level of water stress',
                                                                      'Life expectancy at birth, total (years)',
                                                                      'Military expenditure (% of GDP)', 'Population density',
                                                                      'Population growth (annual %)', '% of total population below 5 meters',
                                                                      'Population living in slums (% of urban population)',
                                                                      'Population, total', 'Poverty $2.15 a day (% of population)',
                                                                      'Poverty $3.65 a day (% of population)',
                                                                      'Poverty $6.85 a day (% of population)', '%  Poverty headcount ratio ',
                                                                      'Time to start a business (days)',
                                                                      'Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)',
                                                                      'Total debt service (% of GNI)',
                                                                      'Total debt service (% of exports of goods, services and primary income)',
                                                                      'Unemployment advanced education', 'Unemployment basic education ',
                                                                      'Unemployment intermediate education',
                                                                      'Unemployment, total (% of total labor force) (modeled ILO estimate)',
                                                                      'Corruption index 2023', 'Global_Peace_Index_2023', 'immigration_100k'
    

                                                                      
                                                                      ]])

In [8]:
# Create a DataFrame with the scaled data
indicators_scaled_df = pd.DataFrame(indicators_scaled, columns=[
                                                               
                                                               'clean fuels for cooking (% of population)',
                                                               'Access to electricity (% of population)', 'Agricultural land (% )',
                                                               'Armed forces (% labor force)', 'precipitation (mm per year)',
                                                               'Birth rate, crude (per 1,000 people)',
                                                               'Central government debt, total (% of GDP)',
                                                               'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
                                                               'Cost of business start-up (% of GNI per capita)',
                                                               'Current account balance (% of GDP)',
                                                               'Current health expenditure (% of GDP)',
                                                               'Death rate, crude (per 1,000 people)',
                                                               'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
                                                               'doing business score', 'Fertility rate, total (births per woman)',
                                                               '% Firms using banks', 'GDP growth (annual %)',
                                                               'GDP per capita (constant 2015 US$)',
                                                               'GDP per capita, PPP (constant 2021 international $)',
                                                               'GNI per capita (2015 US$)', 'government  consumption exp (% of GDP)',
                                                               'Government expenditure on education, total (% of GDP)',
                                                               'Human Capital Index',
                                                               'Industry (including construction), value added (% of GDP)',
                                                               'Inflation, consumer prices (annual %)', 'Level of water stress',
                                                               'Life expectancy at birth, total (years)',
                                                               'Military expenditure (% of GDP)', 'Population density',
                                                               'Population growth (annual %)', '% of total population below 5 meters',
                                                               'Population living in slums (% of urban population)',
                                                               'Population, total', 'Poverty $2.15 a day (% of population)',
                                                               'Poverty $3.65 a day (% of population)',
                                                               'Poverty $6.85 a day (% of population)', '%  Poverty headcount ratio ',
                                                               'Time to start a business (days)',
                                                               'Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)',
                                                               'Total debt service (% of GNI)',
                                                               'Total debt service (% of exports of goods, services and primary income)',
                                                               'Unemployment advanced education', 'Unemployment basic education ',
                                                               'Unemployment intermediate education',
                                                               'Unemployment, total (% of total labor force) (modeled ILO estimate)',
                                                               'Corruption index 2023', 'Global_Peace_Index_2023', 'immigration_100k'
                                                        
                                                               ])

# # Copy the crypto names from the original DataFrame

indicators_scaled_df['country-year'] = full_indicators_df['country'].astype(str) + ' - ' + full_indicators_df['year'].astype(str)



# Display the scaled DataFrame
indicators_scaled_df.head()

clean fuels for cooking (% of population)  \
0                                  -0.952155   
1                                  -0.921814   
2                                  -0.883888   
3                                  -0.856076   
4                                  -0.825735   

   Access to electricity (% of population)  Agricultural land (% )  \
0                                -0.449773                0.887050   
1                                 0.517416                0.887050   
2                                 0.517416                0.887050   
3                                 0.358678                0.893859   
4                                 0.517416                0.893859   

   Armed forces (% labor force)  precipitation (mm per year)  \
0                      1.668181                    -0.937583   
1                      1.610880                    -0.937583   
2                      1.524272                    -0.937583   
3                      1.175262                    -0.937583   
4                      1.215508                    -0.937583   

   Birth rate, crude (per 1,000 people)  \
0                              1.911776   
1                              1.826180   
2                              1.767536   
3                              1.726564   
4                              1.681050   

   Central government debt, total (% of GDP)  \
0                                  -0.599072   
1                                  -0.599072   
2                                  -0.599072   
3                                  -0.599072   
4                                  -0.599072   

   Consumer price index (2010 = 100)  Control of Corruption: Estimate  \
0                          -0.098180                        -1.247837   
1                          -0.095043                        -1.430819   
2                          -0.091327                        -1.420804   
3                          -0.090836                        -1.393977   
4                          -0.089020                        -1.311552   

   Cost of business start-up (% of GNI per capita)  ...  \
0                                        -0.071469  ...   
1                                        -0.048522  ...   
2                                         1.542467  ...   
3                                        -0.392726  ...   
4                                        -0.382527  ...   

   Total debt service (% of GNI)  \
0                      -0.538305   
1                      -0.532740   
2                      -0.531284   
3                      -0.529327   
4                      -0.538452   

   Total debt service (% of exports of goods, services and primary income)  \
0                                          -0.477919                         
1                                          -0.404039                         
2                                          -0.402597                         
3                                          -0.476190                         
4                                          -0.501171                         

   Unemployment advanced education  Unemployment basic education   \
0                        -0.952657                      -0.898684   
1                        -0.952657                      -0.898684   
2                         1.688640                       0.696312   
3                         1.688640                       0.696312   
4                         1.688640                       0.696312   

   Unemployment intermediate education  \
0                            -1.019808   
1                            -1.019808   
2                             1.421518   
3                             1.421518   
4                             1.421518   

   Unemployment, total (% of total labor force) (modeled ILO estimate)  \
0                                           0.340125                     
1                                           0.527347            

In [9]:
## Rename the indicators to be more ... friendly

indicators_scaled_df = indicators_scaled_df.rename(columns={


                                            'Access to clean fuels and technologies for cooking (% of population)'    :   'clean fuels for cooking (% of population)'  ,
                                            'Access to electricity (% of population)'    :   'Access to electricity (% of population)'  ,
                                            'Agricultural land (% of land area)'    :   'Agricultural land (% )'  ,
                                            'Armed forces personnel (% of total labor force)'    :   'Armed forces (% labor force)'  ,
                                            'Average precipitation in depth (mm per year)'    :   'precipitation (mm per year)'  ,
                                            'Birth rate crude (per 1000 people)'    :   'Birth rate crude'  ,
                                            'Central government debt total (% of GDP)'    :   'government debt (% of GDP)'  ,
                                            'Consumer price index (2010 = 100) Control of Corruption: Estimate'    :   'Consumer price index (2010)'  ,
                                            'Cost of business start-up procedures (% of GNI per capita)'    :   'Cost of business start-up (% of GNI per capita)'  ,
                                            'Current account balance (% of GDP)'    :   'Current account balance (% of GDP)'  ,
                                            'Current health expenditure (% of GDP)'    :   'Current health expenditure (% of GDP)'  ,
                                            'Death rate crude (per 1000 people)'    :   'Death rate crude (per 1000 people)'  ,
                                            'Debt service (PPG and IMF only % of exports of goods services and primary income)'    :   'Debt service'  ,
                                            'Ease of doing business score (0 = lowest performance to 100 = best performance)'    :   'doing business score'  ,
                                            'Fertility rate total (births per woman)'    :   'Fertility rate per woman'  ,
                                            'Firms using banks to finance investment (% of firms)'    :   '% Firms using banks'  ,
                                            'GDP growth (annual %) GDP per capita (constant 2015 US$)'    :   'GDP per capita growth '  ,
                                            'GDP per capita PPP (constant 2021 international $)'    :   'GDP per capita PPP ( 2021)'  ,
                                            'GNI per capita (constant 2015 US$)'    :   'GNI per capita (2015 US$)'  ,
                                            'General government final consumption expenditure (% of GDP)'    :   'government  consumption exp (% of GDP)'  ,
                                            'Government expenditure on education total (% of GDP)'    :   'Government expenditure on education total (% of GDP)'  ,
                                            'Human Capital Index (HCI) (scale 0-1)'    :   'Human Capital Index'  ,
                                            'Industry (including construction) value added (% of GDP)'    :   'Industry (% of GDP)'  ,
                                            'Inflation consumer prices (annual %)'    :   'Inflation consumer prices (%)'  ,
                                            'Level of water stress: freshwater withdrawal as a proportion of available freshwater resources'    :   'Level of water stress'  ,
                                            'Life expectancy at birth total (years)'    :   'Life expectancy at birth total (years)'  ,
                                            'Military expenditure (% of GDP)'    :   'Military expenditure (% of GDP)'  ,
                                            'Population density (people per sq. km of land area)'    :   'Population density'  ,
                                            'Population growth (annual %)'    :   'Population growth (annual %)'  ,
                                            'Population living in areas where elevation is below 5 meters (% of total population)'    :   '% of total population below 5 meters'  ,
                                            'Population living in slums (% of urban population)'    :   'Population living in slums (% of urban population)'  ,
                                            'Population total'    :   'Population total'  ,
                                            'Poverty headcount ratio at $2.15 a day (2017 PPP) (% of population)'    :   'Poverty $2.15 a day (% of population)'  ,
                                            'Poverty headcount ratio at $3.65 a day (2017 PPP) (% of population)'    :   'Poverty $3.65 a day (% of population)'  ,
                                            'Poverty headcount ratio at $6.85 a day (2017 PPP) (% of population)'    :   'Poverty $6.85 a day (% of population)'  ,
                                            'Poverty headcount ratio at national poverty lines (% of population)'    :   '%  Poverty headcount ratio '  ,
                                            'Time required to start a business (days)'    :   'Time to start a business (days)'  ,
                                            'Total alcohol consumption per capita (liters of pure alcohol projected estimates 15+ years of age)'    :   'Alcohol consumption per capita'  ,
                                            'Total debt service (% of GNI)'    :   'Total debt service (% of GNI)'  ,
                                            'Total debt service (% of exports of goods services and primary income)'    :   'Total debt service (% of exports)'  ,
                                            'Unemployment with advanced education (% of total labor force with advanced education)'    :   'Unemployment advanced education'  ,
                                            'Unemployment with basic education (% of total labor force with basic education)'    :   'Unemployment basic education '  ,
                                            'Unemployment with intermediate education (% of total labor force with intermediate education)'    :   'Unemployment intermediate education'  ,
                                            'Unemployment total (% of total labor force) (modeled ILO estimate)]'    :   'Unemployment total'  ,
                                            'CPI_score_2023' : 'Corruption index 2023'


                                                    })
indicators_scaled_df.head(5)

clean fuels for cooking (% of population)  \
0                                  -0.952155   
1                                  -0.921814   
2                                  -0.883888   
3                                  -0.856076   
4                                  -0.825735   

   Access to electricity (% of population)  Agricultural land (% )  \
0                                -0.449773                0.887050   
1                                 0.517416                0.887050   
2                                 0.517416                0.887050   
3                                 0.358678                0.893859   
4                                 0.517416                0.893859   

   Armed forces (% labor force)  precipitation (mm per year)  \
0                      1.668181                    -0.937583   
1                      1.610880                    -0.937583   
2                      1.524272                    -0.937583   
3                      1.175262                    -0.937583   
4                      1.215508                    -0.937583   

   Birth rate, crude (per 1,000 people)  \
0                              1.911776   
1                              1.826180   
2                              1.767536   
3                              1.726564   
4                              1.681050   

   Central government debt, total (% of GDP)  \
0                                  -0.599072   
1                                  -0.599072   
2                                  -0.599072   
3                                  -0.599072   
4                                  -0.599072   

   Consumer price index (2010 = 100)  Control of Corruption: Estimate  \
0                          -0.098180                        -1.247837   
1                          -0.095043                        -1.430819   
2                          -0.091327                        -1.420804   
3                          -0.090836                        -1.393977   
4                          -0.089020                        -1.311552   

   Cost of business start-up (% of GNI per capita)  ...  \
0                                        -0.071469  ...   
1                                        -0.048522  ...   
2                                         1.542467  ...   
3                                        -0.392726  ...   
4                                        -0.382527  ...   

   Total debt service (% of GNI)  \
0                      -0.538305   
1                      -0.532740   
2                      -0.531284   
3                      -0.529327   
4                      -0.538452   

   Total debt service (% of exports of goods, services and primary income)  \
0                                          -0.477919                         
1                                          -0.404039                         
2                                          -0.402597                         
3                                          -0.476190                         
4                                          -0.501171                         

   Unemployment advanced education  Unemployment basic education   \
0                        -0.952657                      -0.898684   
1                        -0.952657                      -0.898684   
2                         1.688640                       0.696312   
3                         1.688640                       0.696312   
4                         1.688640                       0.696312   

   Unemployment intermediate education  \
0                            -1.019808   
1                            -1.019808   
2                             1.421518   
3                             1.421518   
4                             1.421518   

   Unemployment, total (% of total labor force) (modeled ILO estimate)  \
0                                           0.340125                     
1                                           0.527347            

In [10]:
indicators_scaled_df.columns

Index(['clean fuels for cooking (% of population)',
       'Access to electricity (% of population)', 'Agricultural land (% )',
       'Armed forces (% labor force)', 'precipitation (mm per year)',
       'Birth rate, crude (per 1,000 people)',
       'Central government debt, total (% of GDP)',
       'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
       'Cost of business start-up (% of GNI per capita)',
       'Current account balance (% of GDP)',
       'Current health expenditure (% of GDP)',
       'Death rate, crude (per 1,000 people)',
       'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
       'doing business score', 'Fertility rate, total (births per woman)',
       '% Firms using banks', 'GDP growth (annual %)',
       'GDP per capita (constant 2015 US$)',
       'GDP per capita, PPP (constant 2021 international $)',
       'GNI per capita (2015 US$)', 'government  consumption exp (% of GDP)',
       'Governmen

In [12]:
# Organize columns

indicators_scaled_df = indicators_scaled_df[[

                                                'country-year', 
                                                'clean fuels for cooking (% of population)',
                                                'Access to electricity (% of population)', 'Agricultural land (% )',
                                            'Armed forces (% labor force)', 'precipitation (mm per year)',
                                            'Birth rate, crude (per 1,000 people)',
                                            'Central government debt, total (% of GDP)',
                                            'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
                                            'Cost of business start-up (% of GNI per capita)',
                                            'Current account balance (% of GDP)',
                                            'Current health expenditure (% of GDP)',
                                            'Death rate, crude (per 1,000 people)',
                                            'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
                                            'doing business score', 'Fertility rate, total (births per woman)',
                                            '% Firms using banks', 'GDP growth (annual %)',
                                            'GDP per capita (constant 2015 US$)',
                                            'GDP per capita, PPP (constant 2021 international $)',
                                            'GNI per capita (2015 US$)', 'government  consumption exp (% of GDP)',
                                            'Government expenditure on education, total (% of GDP)',
                                            'Human Capital Index',
                                            'Industry (including construction), value added (% of GDP)',
                                            'Inflation, consumer prices (annual %)', 'Level of water stress',
                                            'Life expectancy at birth, total (years)',
                                            'Military expenditure (% of GDP)', 'Population density',
                                            'Population growth (annual %)', '% of total population below 5 meters',
                                            'Population living in slums (% of urban population)',
                                            'Population, total', 'Poverty $2.15 a day (% of population)',
                                            'Poverty $3.65 a day (% of population)',
                                            'Poverty $6.85 a day (% of population)', '%  Poverty headcount ratio ',
                                            'Time to start a business (days)',
                                            'Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)',
                                            'Total debt service (% of GNI)',
                                            'Total debt service (% of exports of goods, services and primary income)',
                                            'Unemployment advanced education', 'Unemployment basic education ',
                                            'Unemployment intermediate education',
                                            'Unemployment, total (% of total labor force) (modeled ILO estimate)',
                                            'Corruption index 2023', 'Global_Peace_Index_2023', 'immigration_100k'
                                                                
                                            ]]
indicators_scaled_df.head(5)

country-year  clean fuels for cooking (% of population)  \
0  Afghanistan - 2015                                  -0.952155   
1  Afghanistan - 2016                                  -0.921814   
2  Afghanistan - 2017                                  -0.883888   
3  Afghanistan - 2018                                  -0.856076   
4  Afghanistan - 2019                                  -0.825735   

   Access to electricity (% of population)  Agricultural land (% )  \
0                                -0.449773                0.887050   
1                                 0.517416                0.887050   
2                                 0.517416                0.887050   
3                                 0.358678                0.893859   
4                                 0.517416                0.893859   

   Armed forces (% labor force)  precipitation (mm per year)  \
0                      1.668181                    -0.937583   
1                      1.610880                    -0.937583   
2                      1.524272                    -0.937583   
3                      1.175262                    -0.937583   
4                      1.215508                    -0.937583   

   Birth rate, crude (per 1,000 people)  \
0                              1.911776   
1                              1.826180   
2                              1.767536   
3                              1.726564   
4                              1.681050   

   Central government debt, total (% of GDP)  \
0                                  -0.599072   
1                                  -0.599072   
2                                  -0.599072   
3                                  -0.599072   
4                                  -0.599072   

   Consumer price index (2010 = 100)  Control of Corruption: Estimate  ...  \
0                          -0.098180                        -1.247837  ...   
1                          -0.095043                        -1.430819  ...   
2                          -0.091327                        -1.420804  ...   
3                          -0.090836                        -1.393977  ...   
4                          -0.089020                        -1.311552  ...   

   Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)  \
0                                          -1.263143                                                      
1                                          -1.262154                                                      
2                                          -1.261165                                                      
3                                          -1.261413                                                      
4                                          -1.261413                                                      

   Total debt service (% of GNI)  \
0                      -0.538305   
1                      -0.532740   
2                      -0.531284   
3                      -0.529327   
4                      -0.538452   

   Total debt service (% of exports of goods, services and primary income)  \
0                                          -0.477919                         
1                                          -0.404039                         
2                                          -0.402597                         
3                                          -0.476190                         
4                                          -0.501171                         

   Unemployment advanced education  Unemployment basic education   \
0                        -0.952657                      -0.898684   
1                        -0.952657                      -0.898684   
2                         1.688640                       0.696312   
3                         1.688640                       0.696312   
4                         1.688640                       0.696312   

   Unemployment intermediate educ

In [13]:
indicators_scaled_df = indicators_scaled_df.set_index(['country-year'])
indicators_scaled_df.head(5)

clean fuels for cooking (% of population)  \
country-year                                                    
Afghanistan - 2015                                  -0.952155   
Afghanistan - 2016                                  -0.921814   
Afghanistan - 2017                                  -0.883888   
Afghanistan - 2018                                  -0.856076   
Afghanistan - 2019                                  -0.825735   

                    Access to electricity (% of population)  \
country-year                                                  
Afghanistan - 2015                                -0.449773   
Afghanistan - 2016                                 0.517416   
Afghanistan - 2017                                 0.517416   
Afghanistan - 2018                                 0.358678   
Afghanistan - 2019                                 0.517416   

                    Agricultural land (% )  Armed forces (% labor force)  \
country-year                                                               
Afghanistan - 2015                0.887050                      1.668181   
Afghanistan - 2016                0.887050                      1.610880   
Afghanistan - 2017                0.887050                      1.524272   
Afghanistan - 2018                0.893859                      1.175262   
Afghanistan - 2019                0.893859                      1.215508   

                    precipitation (mm per year)  \
country-year                                      
Afghanistan - 2015                    -0.937583   
Afghanistan - 2016                    -0.937583   
Afghanistan - 2017                    -0.937583   
Afghanistan - 2018                    -0.937583   
Afghanistan - 2019                    -0.937583   

                    Birth rate, crude (per 1,000 people)  \
country-year                                               
Afghanistan - 2015                              1.911776   
Afghanistan - 2016                              1.826180   
Afghanistan - 2017                              1.767536   
Afghanistan - 2018                              1.726564   
Afghanistan - 2019                              1.681050   

                    Central government debt, total (% of GDP)  \
country-year                                                    
Afghanistan - 2015                                  -0.599072   
Afghanistan - 2016                                  -0.599072   
Afghanistan - 2017                                  -0.599072   
Afghanistan - 2018                                  -0.599072   
Afghanistan - 2019                                  -0.599072   

                    Consumer price index (2010 = 100)  \
country-year                                            
Afghanistan - 2015                          -0.098180   
Afghanistan - 2016                          -0.095043   
Afghanistan - 2017                          -0.091327   
Afghanistan - 2018                          -0.090836   
Afghanistan - 2019                          -0.089020   

                    Control of Corruption: Estimate  \
country-year                                          
Afghanistan - 2015                        -1.247837   
Afghanistan - 2016                        -1.430819   
Afghanistan - 2017                        -1.420804   
Afghanistan - 2018                        -1.393977   
Afghanistan - 2019                        -1.311552   

                    Cost of business start-up (% of GNI per capita)  ...  \
country-year                                                         ...   
Afghanistan - 2015                                        -0.071469  ...   
Afghanistan - 2016                                        -0.048522  ...   
Afghanistan - 2017                                         1.542467  ...   
Afghanistan - 2018                                        -0.392726  ...   
Afghanistan - 2019                                        -0.382527  ...   

                    Total alcohol c

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [14]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,12))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [15]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(indicators_scaled_df)
    inertia.append(k_model.inertia_)


In [16]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
df_elbow.head(12)

k       inertia
0    1  81487.000000
1    2  69539.341081
2    3  67894.214661
3    4  62578.421353
4    5  59612.120303
5    6  55812.891482
6    7  53479.023491
7    8  53194.046248
8    9  50659.346194
9   10  48394.421218
10  11  46574.198183

In [17]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_original_data = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve without PCA", 
    xticks=k
)

elbow_original_data


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  4

---

### Cluster the immigration data with K-means Using the Original Scaled DataFrame

In [18]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [19]:
# Fit the K-Means model using the scaled DataFrame
model.fit(indicators_scaled_df)

KMeans(n_clusters=4, random_state=0)

In [22]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
kmeans_predictions = model.predict(indicators_scaled_df)

# Print the resulting array of cluster values.
print(kmeans_predictions)

[1 1 1 ... 1 3 3]


In [23]:
# Create a copy of the scaled DataFrame
indicators_prediction_df = indicators_scaled_df.copy()

In [24]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
indicators_prediction_df['countries_cluster_original_data'] = kmeans_predictions

# Display the copy of the scaled DataFrame
indicators_prediction_df.head()

clean fuels for cooking (% of population)  \
country-year                                                    
Afghanistan - 2015                                  -0.952155   
Afghanistan - 2016                                  -0.921814   
Afghanistan - 2017                                  -0.883888   
Afghanistan - 2018                                  -0.856076   
Afghanistan - 2019                                  -0.825735   

                    Access to electricity (% of population)  \
country-year                                                  
Afghanistan - 2015                                -0.449773   
Afghanistan - 2016                                 0.517416   
Afghanistan - 2017                                 0.517416   
Afghanistan - 2018                                 0.358678   
Afghanistan - 2019                                 0.517416   

                    Agricultural land (% )  Armed forces (% labor force)  \
country-year                                                               
Afghanistan - 2015                0.887050                      1.668181   
Afghanistan - 2016                0.887050                      1.610880   
Afghanistan - 2017                0.887050                      1.524272   
Afghanistan - 2018                0.893859                      1.175262   
Afghanistan - 2019                0.893859                      1.215508   

                    precipitation (mm per year)  \
country-year                                      
Afghanistan - 2015                    -0.937583   
Afghanistan - 2016                    -0.937583   
Afghanistan - 2017                    -0.937583   
Afghanistan - 2018                    -0.937583   
Afghanistan - 2019                    -0.937583   

                    Birth rate, crude (per 1,000 people)  \
country-year                                               
Afghanistan - 2015                              1.911776   
Afghanistan - 2016                              1.826180   
Afghanistan - 2017                              1.767536   
Afghanistan - 2018                              1.726564   
Afghanistan - 2019                              1.681050   

                    Central government debt, total (% of GDP)  \
country-year                                                    
Afghanistan - 2015                                  -0.599072   
Afghanistan - 2016                                  -0.599072   
Afghanistan - 2017                                  -0.599072   
Afghanistan - 2018                                  -0.599072   
Afghanistan - 2019                                  -0.599072   

                    Consumer price index (2010 = 100)  \
country-year                                            
Afghanistan - 2015                          -0.098180   
Afghanistan - 2016                          -0.095043   
Afghanistan - 2017                          -0.091327   
Afghanistan - 2018                          -0.090836   
Afghanistan - 2019                          -0.089020   

                    Control of Corruption: Estimate  \
country-year                                          
Afghanistan - 2015                        -1.247837   
Afghanistan - 2016                        -1.430819   
Afghanistan - 2017                        -1.420804   
Afghanistan - 2018                        -1.393977   
Afghanistan - 2019                        -1.311552   

                    Cost of business start-up (% of GNI per capita)  ...  \
country-year                                                         ...   
Afghanistan - 2015                                        -0.071469  ...   
Afghanistan - 2016                                        -0.048522  ...   
Afghanistan - 2017                                         1.542467  ...   
Afghanistan - 2018                                        -0.392726  ...   
Afghanistan - 2019                                        -0.382527  ...   

                    Total debt serv

In [26]:
# Create a scatter plot using hvPlot by setting
    # `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
    # Color the graph points with the labels found using K-Means and
    # add the crypto name in the `hover_cols` parameter to identify
    # the cryptocurrency represented by each data point.

clusters_original_data = indicators_prediction_df.hvplot.scatter(
    x="immigration_100k", 
    y="Global_Peace_Index_2023", 
    by="countries_cluster_original_data",
    title = "Clusters with original data",
    hover_cols = 'country-year'
)

clusters_original_data

:NdOverlay   [countries_cluster_original_data]
   :Scatter   [immigration_100k]   (Global_Peace_Index_2023,country-year)

---

### Optimize Clusters with Principal Component Analysis.

In [27]:
# Create a PCA model instance and set `n_components=5`.
pca = PCA(n_components=5)

In [28]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to four principal components.
indicators_pca = pca.fit_transform(indicators_scaled_df)

# View the scaled PCA data
indicators_pca[0:5]

array([[ 3.27020581, -1.97997571, -0.99661033,  0.75322919, -0.33903367],
       [ 3.92600645, -0.82085159, -0.01775231,  0.19169539, -0.76194156],
       [ 4.00193096,  1.25005058, -0.78806944,  0.37849199, -2.52465574],
       [ 3.54891034,  1.61980276, -0.49000297,  0.4737365 , -2.10725466],
       [ 3.41064539,  1.62411933, -0.45960017,  0.55350187, -2.06732118]])

In [29]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.20773766, 0.08690068, 0.06621939, 0.04976276, 0.04811346])

In [30]:
# Use 'sum' on the explained varaibles array to obtain the total explained variance

total_explained_variance = sum(pca.explained_variance_ratio_)
print(f'The total explained variance is : {total_explained_variance}')

The total explained variance is : 0.45873395541721945


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 45.9%

In [32]:
# Create a new DataFrame with the PCA data.
indicators_pca_df = pd.DataFrame(
    indicators_pca,
    columns=["PCA1", "PCA2", 'PCA3', 'PCA4', 'PCA5']
)

# Copy the crypto names from the original scaled DataFrame
indicators_pca_df['country-year'] = full_indicators_df['country'].astype(str) + ' - ' + full_indicators_df['year'].astype(str)

# # Set the coin_id column as index
indicators_pca_df = indicators_pca_df.set_index(['country-year'])

# Display the scaled PCA DataFrame
indicators_pca_df.head(5)

PCA1      PCA2      PCA3      PCA4      PCA5
country-year                                                        
Afghanistan - 2015  3.270206 -1.979976 -0.996610  0.753229 -0.339034
Afghanistan - 2016  3.926006 -0.820852 -0.017752  0.191695 -0.761942
Afghanistan - 2017  4.001931  1.250051 -0.788069  0.378492 -2.524656
Afghanistan - 2018  3.548910  1.619803 -0.490003  0.473736 -2.107255
Afghanistan - 2019  3.410645  1.624119 -0.459600  0.553502 -2.067321

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [33]:
# Create a list with the number of k-values from 1 to 11
    # we already have the list
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [34]:
# Create an empty list to store the inertia values

inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model_pca = KMeans(n_clusters=i, random_state=1)
    k_model_pca.fit(indicators_pca_df)
    inertia_pca.append(k_model_pca.inertia_)

In [35]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k": k, "inertia": inertia_pca}

# Create a DataFrame with the data to plot the Elbow curve

df_elbow_pca = pd.DataFrame(elbow_data_pca)
df_elbow_pca.head()


k       inertia
0  1  37380.853825
1  2  25532.164918
2  3  20374.028544
3  4  16985.304591
4  5  14900.773142

In [37]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.

elbow_with_pca = df_elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve_with_PCA", 
    xticks=k
)

elbow_with_pca

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No

### Cluster Immigration data with K-means Using the Scaled PCA DataFrame

In [38]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [40]:
# Fit the K-Means model using the PCA data
model.fit(indicators_pca_df)

KMeans(n_clusters=4, random_state=0)

In [41]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
kmeans_predictions_pca = model.predict(indicators_pca_df)

# Print the resulting array of cluster values.
print(kmeans_predictions_pca)

[3 3 2 ... 2 2 2]


In [42]:
# Create a copy of the scaled PCA DataFrame

indicators_pca_new_df = indicators_pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
indicators_pca_new_df['countries_cluster_withPCA'] = kmeans_predictions_pca

# Display the copy of the scaled PCA DataFrame
indicators_pca_new_df.head(5)

PCA1      PCA2      PCA3      PCA4      PCA5  \
country-year                                                           
Afghanistan - 2015  3.270206 -1.979976 -0.996610  0.753229 -0.339034   
Afghanistan - 2016  3.926006 -0.820852 -0.017752  0.191695 -0.761942   
Afghanistan - 2017  4.001931  1.250051 -0.788069  0.378492 -2.524656   
Afghanistan - 2018  3.548910  1.619803 -0.490003  0.473736 -2.107255   
Afghanistan - 2019  3.410645  1.624119 -0.459600  0.553502 -2.067321   

                    countries_cluster_withPCA  
country-year                                   
Afghanistan - 2015                          3  
Afghanistan - 2016                          3  
Afghanistan - 2017                          2  
Afghanistan - 2018                          2  
Afghanistan - 2019                          2

In [43]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
clusters_with_PCA = indicators_pca_new_df.hvplot.scatter(
    x="PCA1", 
    y="PCA2", 
    by="countries_cluster_withPCA",
    title = "Clusters with PCA" ,
    hover_cols = 'country-year'
)

clusters_with_PCA

:NdOverlay   [countries_cluster_withPCA]
   :Scatter   [PCA1]   (PCA2,country-year)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [44]:
# Composite plot to contrast the Elbow curves

plots = (elbow_original_data + elbow_with_pca).cols(1)
plots

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [49]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!


plots = (clusters_original_data + clusters_with_PCA).cols(1)
plots

:Layout
   .NdOverlay.I  :NdOverlay   [countries_cluster_original_data]
      :Scatter   [immigration_100k]   (Global_Peace_Index_2023,country-year)
   .NdOverlay.II :NdOverlay   [countries_cluster_withPCA]
      :Scatter   [PCA1]   (PCA2,country-year)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 

  1)  regarding the number of clusters
  the fact that the optimal k was the same with and without use of PCA suggests the following : 
  - PCA did not significantly distort the structure of the data relevant to clustering..
  - The original dataset likely had a strong cluster structure and the clustering information was present in the original feature set.
  - the clusters in the original dataset are well separated, so the PCA did not change significantly the separability.

On other hand this consistancy is positive because 
  - it gives more confidence in the clustering results


  2) regarding the clustering assignements
  we observe that both methods (with and without PCA) give the same clusters (whatever is the # of the cluster)
  all the countries belong to the same cluster in both methods as shown in the dataframe below

  this reinforce the conclusions above

In [47]:
# to have the detailled content of the clusters, we will merge the 2 dataframes (with original data and with PCA)
# than we will compare the content of each cluster
# no need to make further queries, the data is limited and allows us to conclude

clusters = pd.merge(indicators_pca_new_df, indicators_prediction_df, on='country-year')
clusters = clusters[['countries_cluster_original_data', 'countries_cluster_withPCA']].sort_values(['countries_cluster_withPCA'])

# clusters.head(5)
clusters

countries_cluster_original_data  \
country-year                                                   
Kuwait - 2018                                              0   
Hong Kong SAR, China - 2020                                0   
Hong Kong SAR, China - 2021                                0   
Hong Kong SAR, China - 2022                                0   
Hong Kong SAR, China - 2023                                0   
...                                                      ...   
Somalia - 2019                                             1   
Somalia - 2018                                             1   
Somalia - 2017                                             1   
Guatemala - 2015                                           2   
Afghanistan - 2015                                         1   

                             countries_cluster_withPCA  
country-year                                            
Kuwait - 2018                                        0  
Hong Kong SAR, China - 2020                          0  
Hong Kong SAR, China - 2021                          0  
Hong Kong SAR, China - 2022                          0  
Hong Kong SAR, China - 2023                          0  
...                                                ...  
Somalia - 2019                                       3  
Somalia - 2018                                       3  
Somalia - 2017                                       3  
Guatemala - 2015                                     3  
Afghanistan - 2015                                   3  

[1663 rows x 2 columns]

In [48]:
clusters.to_csv('Output_analysis/Clusters_analysis.csv')

### Mearge cluster information in full indicators dataframe

In [50]:
indicators_scaled_df.head(1)


clean fuels for cooking (% of population)  \
country-year                                                    
Afghanistan - 2015                                  -0.952155   

                    Access to electricity (% of population)  \
country-year                                                  
Afghanistan - 2015                                -0.449773   

                    Agricultural land (% )  Armed forces (% labor force)  \
country-year                                                               
Afghanistan - 2015                 0.88705                      1.668181   

                    precipitation (mm per year)  \
country-year                                      
Afghanistan - 2015                    -0.937583   

                    Birth rate, crude (per 1,000 people)  \
country-year                                               
Afghanistan - 2015                              1.911776   

                    Central government debt, total (% of GDP)  \
country-year                                                    
Afghanistan - 2015                                  -0.599072   

                    Consumer price index (2010 = 100)  \
country-year                                            
Afghanistan - 2015                           -0.09818   

                    Control of Corruption: Estimate  \
country-year                                          
Afghanistan - 2015                        -1.247837   

                    Cost of business start-up (% of GNI per capita)  ...  \
country-year                                                         ...   
Afghanistan - 2015                                        -0.071469  ...   

                    Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)  \
country-year                                                                                                               
Afghanistan - 2015                                          -1.263143                                                      

                    Total debt service (% of GNI)  \
country-year                                        
Afghanistan - 2015                      -0.538305   

                    Total debt service (% of exports of goods, services and primary income)  \
country-year                                                                                  
Afghanistan - 2015                                          -0.477919                         

                    Unemployment advanced education  \
country-year                                          
Afghanistan - 2015                        -0.952657   

                    Unemployment basic education   \
country-year                                        
Afghanistan - 2015                      -0.898684   

                    Unemployment intermediate education  \
country-year                                              
Afghanistan - 2015                            -1.019808   

                    Unemployment, total (% of total labor force) (modeled ILO estimate)  \
country-year                                                                              
Afghanistan - 2015                                           0.340125                     

                    Corruption index 2023  Global_Peace_Index_2023  \
country-year                                                         
Afghanistan - 2015              -0.958168                 1.626599   

                    immigration_100k  
country-year                          
Afghanistan - 2015         -0.099321  

[1 rows x 49 columns]

In [51]:
# Reset the index, moving 'country-year' back into a column
indicators_scaled_df.reset_index(inplace=True)
indicators_scaled_df.head(1)

country-year  clean fuels for cooking (% of population)  \
0  Afghanistan - 2015                                  -0.952155   

   Access to electricity (% of population)  Agricultural land (% )  \
0                                -0.449773                 0.88705   

   Armed forces (% labor force)  precipitation (mm per year)  \
0                      1.668181                    -0.937583   

   Birth rate, crude (per 1,000 people)  \
0                              1.911776   

   Central government debt, total (% of GDP)  \
0                                  -0.599072   

   Consumer price index (2010 = 100)  Control of Corruption: Estimate  ...  \
0                           -0.09818                        -1.247837  ...   

   Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)  \
0                                          -1.263143                                                      

   Total debt service (% of GNI)  \
0                      -0.538305   

   Total debt service (% of exports of goods, services and primary income)  \
0                                          -0.477919                         

   Unemployment advanced education  Unemployment basic education   \
0                        -0.952657                      -0.898684   

   Unemployment intermediate education  \
0                            -1.019808   

   Unemployment, total (% of total labor force) (modeled ILO estimate)  \
0                                           0.340125                     

   Corruption index 2023  Global_Peace_Index_2023  immigration_100k  
0              -0.958168                 1.626599         -0.099321  

[1 rows x 50 columns]

In [52]:
new_df = pd.merge(indicators_scaled_df, clusters, on = 'country-year' )
new_df.columns

Index(['country-year', 'clean fuels for cooking (% of population)',
       'Access to electricity (% of population)', 'Agricultural land (% )',
       'Armed forces (% labor force)', 'precipitation (mm per year)',
       'Birth rate, crude (per 1,000 people)',
       'Central government debt, total (% of GDP)',
       'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
       'Cost of business start-up (% of GNI per capita)',
       'Current account balance (% of GDP)',
       'Current health expenditure (% of GDP)',
       'Death rate, crude (per 1,000 people)',
       'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
       'doing business score', 'Fertility rate, total (births per woman)',
       '% Firms using banks', 'GDP growth (annual %)',
       'GDP per capita (constant 2015 US$)',
       'GDP per capita, PPP (constant 2021 international $)',
       'GNI per capita (2015 US$)', 'government  consumption exp (% of GDP)',
 

In [53]:
# to have the population of each cluster
x = new_df.groupby('countries_cluster_withPCA')['country-year'].count()
x

countries_cluster_withPCA
0    501
1    586
2    245
3    331
Name: country-year, dtype: int64

In [56]:
new_df.head(5)

country-year  clean fuels for cooking (% of population)  \
0  Afghanistan - 2015                                  -0.952155   
1  Afghanistan - 2016                                  -0.921814   
2  Afghanistan - 2017                                  -0.883888   
3  Afghanistan - 2018                                  -0.856076   
4  Afghanistan - 2019                                  -0.825735   

   Access to electricity (% of population)  Agricultural land (% )  \
0                                -0.449773                0.887050   
1                                 0.517416                0.887050   
2                                 0.517416                0.887050   
3                                 0.358678                0.893859   
4                                 0.517416                0.893859   

   Armed forces (% labor force)  precipitation (mm per year)  \
0                      1.668181                    -0.937583   
1                      1.610880                    -0.937583   
2                      1.524272                    -0.937583   
3                      1.175262                    -0.937583   
4                      1.215508                    -0.937583   

   Birth rate, crude (per 1,000 people)  \
0                              1.911776   
1                              1.826180   
2                              1.767536   
3                              1.726564   
4                              1.681050   

   Central government debt, total (% of GDP)  \
0                                  -0.599072   
1                                  -0.599072   
2                                  -0.599072   
3                                  -0.599072   
4                                  -0.599072   

   Consumer price index (2010 = 100)  Control of Corruption: Estimate  ...  \
0                          -0.098180                        -1.247837  ...   
1                          -0.095043                        -1.430819  ...   
2                          -0.091327                        -1.420804  ...   
3                          -0.090836                        -1.393977  ...   
4                          -0.089020                        -1.311552  ...   

   Total debt service (% of exports of goods, services and primary income)  \
0                                          -0.477919                         
1                                          -0.404039                         
2                                          -0.402597                         
3                                          -0.476190                         
4                                          -0.501171                         

   Unemployment advanced education  Unemployment basic education   \
0                        -0.952657                      -0.898684   
1                        -0.952657                      -0.898684   
2                         1.688640                       0.696312   
3                         1.688640                       0.696312   
4                         1.688640                       0.696312   

   Unemployment intermediate education  \
0                            -1.019808   
1                            -1.019808   
2                             1.421518   
3                             1.421518   
4                             1.421518   

   Unemployment, total (% of total labor force) (modeled ILO estimate)  \
0                                           0.340125                     
1                                           0.527347                     
2                                           0.709374                     
3                                           0.711452                     
4                                           0.709547                     

   Corruption index 2023  Global_Peace_Index_2023  immigration_100k  \
0              -0.958168                 1.626599         -0.099321   
1              -0.958168                 1.626599    

In [58]:
# Create a dictionary where each key is a cluster label and each value is the corresponding DataFrame
cluster_dfs = {cluster: group.copy() for cluster, group in new_df.groupby('countries_cluster_withPCA')}



In [61]:
df_cluster_0 = cluster_dfs[0]
df_cluster_0

country-year  clean fuels for cooking (% of population)  \
75    Australia - 2015                                    0.87841   
76    Australia - 2016                                    0.87841   
77    Australia - 2017                                    0.87841   
78    Australia - 2018                                    0.87841   
79    Australia - 2019                                    0.87841   
...                ...                                        ...   
1594    Uruguay - 2019                                    0.87841   
1595    Uruguay - 2020                                    0.87841   
1596    Uruguay - 2021                                    0.87841   
1597    Uruguay - 2022                                    0.87841   
1598    Uruguay - 2023                                    0.87841   

      Access to electricity (% of population)  Agricultural land (% )  \
75                                   0.602321                0.318200   
76                                   0.602321                0.283806   
77                                   0.602321                0.452592   
78                                   0.602321                0.377872   
79                                   0.602321                0.398552   
...                                       ...                     ...   
1594                                 0.598630                1.890458   
1595                                 0.602321                1.874244   
1596                                 0.602321                1.875918   
1597                                 0.602321                1.897355   
1598                                 0.602321                1.897355   

      Armed forces (% labor force)  precipitation (mm per year)  \
75                       -0.439682                     -0.67993   
76                       -0.443782                     -0.67993   
77                       -0.449043                     -0.67993   
78                       -0.455116                     -0.67993   
79                       -0.456562                     -0.67993   
...                            ...                          ...   
1594                      0.085734                      0.27351   
1595                      0.114132                      0.27351   
1596                      0.114132                      0.27351   
1597                      0.114132                      0.27351   
1598                      0.114132                      0.27351   

      Birth rate, crude (per 1,000 people)  \
75                               -0.645560   
76                               -0.645560   
77                               -0.675179   
78                               -0.675179   
79                               -0.724542   
...                                    ...   
1594                             -0.850716   
1595                             -0.898401   
1596                             -0.885172   
1597                             -0.890503   
1598                             -0.890503   

      Central government debt, total (% of GDP)  \
75                                     0.565421   
76                                     0.770751   
77                                     0.736640   
78                                     0.766368   
79                                     0.909745   
...                                         ...   
1594                                   0.656848   
1595                                   0.911056   
1596                                   0.911056   
1597                                   0.911056   
1598                                   0.911056   

      Consumer price index (2010 = 100)  Control of Corruption: Estimate  ...  \
75                            -0.109399                         1.904539  ...   
76                            -0.108629                         1.836426  ...   
77                            -0.107438                         1.817412  ...   
78       

In [62]:
df_cluster_1 = cluster_dfs[1]
df_cluster_1

country-year  clean fuels for cooking (% of population)  \
9                Albania - 2015                                   0.286763   
10               Albania - 2016                                   0.329746   
11               Albania - 2017                                   0.362616   
12               Albania - 2018                                   0.390428   
13               Albania - 2019                                   0.423297   
...                         ...                                        ...   
1631  West Bank and Gaza - 2019                                  -1.649995   
1632  West Bank and Gaza - 2020                                  -1.649995   
1633  West Bank and Gaza - 2021                                  -1.649995   
1634  West Bank and Gaza - 2022                                  -1.649995   
1635  West Bank and Gaza - 2023                                  -1.649995   

      Access to electricity (% of population)  Agricultural land (% )  \
9                                    0.602321                0.209097   
10                                   0.598630                0.221091   
11                                   0.598630                0.209067   
12                                   0.602321                0.208743   
13                                   0.602321                0.208611   
...                                       ...                     ...   
1631                                 0.602321                1.443441   
1632                                 0.602321                1.314612   
1633                                 0.602321                1.189291   
1634                                 0.602321                1.189291   
1635                                 0.602321                1.189291   

      Armed forces (% labor force)  precipitation (mm per year)  \
9                        -0.312831                     0.503780   
10                       -0.327565                     0.503780   
11                       -0.310344                     0.503780   
12                       -0.319639                     0.503780   
13                       -0.371779                     0.503780   
...                            ...                          ...   
1631                     -0.736589                    -0.844231   
1632                     -0.736589                    -0.844231   
1633                     -0.736589                    -0.844231   
1634                     -0.736589                    -0.844231   
1635                     -0.736589                    -0.844231   

      Birth rate, crude (per 1,000 people)  \
9                                -0.762256   
10                               -0.804610   
11                               -0.845977   
12                               -0.880828   
13                               -0.898006   
...                                    ...   
1631                              0.989958   
1632                              0.924206   
1633                              0.868524   
1634                              0.814618   
1635                              0.814618   

      Central government debt, total (% of GDP)  \
9                                      1.401735   
10                                     1.423560   
11                                     1.302134   
12                                     0.995116   
13                                     1.275066   
...                                         ...   
1631                                  -0.599072   
1632                                  -0.599072   
1633                                  -0.599072   
1634                                  -0.599072   
1635                                  -0.599072   

      Consumer price index (2010 = 100)  Control of Corruption: Estimate  ...  \
9                             -0.109748                        -0.449321  ...   
10                            -0.108983                        -0.377547

In [64]:
df_cluster_2 = cluster_dfs[2]
df_cluster_2

country-year  clean fuels for cooking (% of population)  \
2     Afghanistan - 2017                                  -0.883888   
3     Afghanistan - 2018                                  -0.856076   
4     Afghanistan - 2019                                  -0.825735   
5     Afghanistan - 2020                                  -0.795394   
6     Afghanistan - 2021                                  -0.767581   
...                  ...                                        ...   
1658     Zimbabwe - 2019                                  -0.883888   
1659     Zimbabwe - 2020                                  -0.878831   
1660     Zimbabwe - 2021                                  -0.878831   
1661     Zimbabwe - 2022                                  -0.871246   
1662     Zimbabwe - 2023                                  -0.871246   

      Access to electricity (% of population)  Agricultural land (% )  \
2                                    0.517416                0.887050   
3                                    0.358678                0.893859   
4                                    0.517416                0.893859   
5                                    0.517416                0.914490   
6                                    0.517416                0.914490   
...                                       ...                     ...   
1658                                -1.365280                0.060801   
1659                                -1.143787                0.071269   
1660                                -1.280374                0.054919   
1661                                -1.239767                0.059510   
1662                                -1.239767                0.059510   

      Armed forces (% labor force)  precipitation (mm per year)  \
2                         1.524272                    -0.937583   
3                         1.175262                    -0.937583   
4                         1.215508                    -0.937583   
5                         0.439168                    -0.937583   
6                         0.439168                    -0.937583   
...                            ...                          ...   
1658                     -0.166245                    -0.526832   
1659                     -0.170225                    -0.526832   
1660                     -0.170225                    -0.526832   
1661                     -0.170225                    -0.526832   
1662                     -0.170225                    -0.526832   

      Birth rate, crude (per 1,000 people)  \
2                                 1.767536   
3                                 1.726564   
4                                 1.681050   
5                                 1.640079   
6                                 1.619444   
...                                    ...   
1658                              1.192547   
1659                              1.142295   
1660                              1.095695   
1661                              1.051169   
1662                              1.051169   

      Central government debt, total (% of GDP)  \
2                                     -0.599072   
3                                     -0.599072   
4                                     -0.599072   
5                                     -0.599072   
6                                     -0.599072   
...                                         ...   
1658                                  -0.599072   
1659                                  -0.599072   
1660                                  -0.599072   
1661                                  -0.599072   
1662                                  -0.599072   

      Consumer price index (2010 = 100)  Control of Corruption: Estimate  ...  \
2                             -0.091327                        -1.420804  ...   
3                             -0.090836                        -1.393977  ...   
4                             -0.089020                        -1.31

In [65]:
df_cluster_3 = cluster_dfs[3]
df_cluster_3

country-year  clean fuels for cooking (% of population)  \
0        Afghanistan - 2015                                  -0.952155   
1        Afghanistan - 2016                                  -0.921814   
18           Algeria - 2015                                   0.868297   
19           Algeria - 2016                                   0.868297   
27    American Samoa - 2015                                  -1.649995   
...                     ...                                        ...   
1642     Yemen, Rep. - 2021                                  -0.395906   
1643     Yemen, Rep. - 2022                                  -0.428775   
1644     Yemen, Rep. - 2023                                  -0.428775   
1654        Zimbabwe - 2015                                  -0.891473   
1655        Zimbabwe - 2016                                  -0.891473   

      Access to electricity (% of population)  Agricultural land (% )  \
0                                   -0.449773                0.887050   
1                                    0.517416                0.887050   
18                                   0.580172               -0.921197   
19                                   0.580172               -0.922990   
27                                  -3.089239               -1.090215   
...                                       ...                     ...   
1642                                -0.324260                0.278444   
1643                                -0.283653                0.278444   
1644                                -0.283653                0.278444   
1654                                -1.845183                0.078548   
1655                                -1.520326                0.088363   

      Armed forces (% labor force)  precipitation (mm per year)  \
0                         1.668181                    -0.937583   
1                         1.610880                    -0.937583   
18                        0.962167                    -1.233822   
19                        0.943468                    -1.233822   
27                       -0.736589                    -1.344600   
...                            ...                          ...   
1642                     -0.361699                    -1.136735   
1643                     -0.361699                    -1.136735   
1644                     -0.361699                    -1.136735   
1654                     -0.124277                    -0.526832   
1655                     -0.134461                    -0.526832   

      Birth rate, crude (per 1,000 people)  \
0                                 1.911776   
1                                 1.826180   
18                                0.589026   
19                                0.565430   
27                               -1.919144   
...                                    ...   
1642                              1.096387   
1643                              1.034682   
1644                              1.034682   
1654                              1.433639   
1655                              1.355743   

      Central government debt, total (% of GDP)  \
0                                     -0.599072   
1                                     -0.599072   
18                                    -0.599072   
19                                    -0.599072   
27                                    -0.599072   
...                                         ...   
1642                                  -0.599072   
1643                                  -0.599072   
1644                                  -0.599072   
1654                                  -0.599072   
1655                                  -0.599072   

      Consumer price index (2010 = 100)  Control of Corruption: Estimate  ...  \
0                             -0.098180                        -1.247837  ...   
1                             -0.095043                        -1.430819  ...   
18                            -0.10

### Define the target and predictors.

In [72]:
# Select features and target
target = 'immigration_100k'
X = df_cluster_0.drop(columns= [target, 'country-year','countries_cluster_original_data','countries_cluster_withPCA'])
y = df_cluster_0[target]

In [73]:
X.shape

(501, 48)

In [74]:
y.shape

(501,)

## Splitting the data

In [75]:
## Splitting the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [76]:
X_train.shape

(400, 48)

In [77]:
X_test.shape

(101, 48)

## Regression Modelling

### --- Linear Regression ---

In [80]:
# --- Linear Regression ---
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)
mse_lr = mean_squared_error(y_test, y_pred_lr)
r2_lr = r2_score(y_test, y_pred_lr)
print("Linear Regression MSE:", mse_lr)
print("Linear Regression R2:", r2_lr)

Linear Regression MSE: 0.054266712759710935
Linear Regression R2: 0.3125801486108165


### --- Random Forest Regression ---

In [81]:
### --- Random Forest Regression ---
rf = RandomForestRegressor(random_state=42)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)
mse_rf = mean_squared_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)
print("Random Forest Regression MSE:", mse_rf)
print("Random Forest Regression R2:", r2_rf)

Random Forest Regression MSE: 0.00906833260423191
Random Forest Regression R2: 0.8851275204608139


The results show that the Random Forest model significantly outperforms the Linear Regression model:

- Mean Squared Error (MSE):

  - Linear Regression: 0.054
  - Random Forest: 0.009
   A lower MSE indicates that the Random Forest model's predictions are, on average, much closer to the actual values.

- R² Score:

  - Linear Regression: 0.31 (approximately 31% of the variance explained)
  - Random Forest: 0.885 (approximately 88.5% of the variance explained)
   The higher R² score of the Random Forest model means it captures a significantly larger proportion of the variability in immigration_100k.
   
In summary, the Random Forest model is more accurate and reliable for predicting immigration per 100k, likely due to its ability to capture non-linear relationships and interactions among the predictors that a simple linear model might miss.